decrypt.go

```diff
@@ -41,6 +66,42 @@ func (p7 *PKCS7) Decrypt(cert *x509.Certificate, pkey crypto.PrivateKey) ([]byte
 	return nil, ErrUnsupportedAlgorithm
 }
 
+// RFC 4055, 4.1
+// The current ASN.1 parser does not support non-integer defaults so the 'default:' tags here do nothing.
+type rsaOAEPAlgorithmParameters struct {
+	HashFunc    pkix.AlgorithmIdentifier `asn1:"optional,explicit,tag:0,default:sha1Identifier"`
+	MaskGenFunc pkix.AlgorithmIdentifier `asn1:"optional,explicit,tag:1,default:mgf1SHA1Identifier"`
+	PSourceFunc pkix.AlgorithmIdentifier `asn1:"optional,explicit,tag:2,default:pSpecifiedEmptyIdentifier"`
+}
+
+func getHashFuncForKeyEncryptionAlgorithm(keyEncryptionAlgorithm pkix.AlgorithmIdentifier) (crypto.Hash, error) {
```

Comment:
I think we should return the maskGenAlgorithm too if available crypto.Hash(0) if not.

Reply:
But the `maskGenAlgorithm` is something different than the `hashAlgorithm`. Currently there's no way to pass the MGF through to where it's used, which is in `rsa.DecryptOAEP`. That function actually reuses the hash that's passed as the `hashAlgorithm`. I understand why you propose to look for the MGF to be set, but I think it's not the right thing to do, until we can provide both hashes to `rsa.DecryptOAEP` independently from each other, so that multiple combinations are possible. If we consider the line that says the MGF should generally be the same as the hash algorithm, that also leads me to believe that the hash algorithm has to be set, and the MGF to be derived from that.

Reply:
Is that not the same as the `MGFHash` in `rsa.OAEPOptions`? I think it can be used with `rsa.PrivateKey.Decrypt`, or am I mistaken?

Reply:
This got me confused for a bit, since I'm still on Go 1.19 at the moment, whereas the `MGFHash` was added only recently: https://github.com/golang/go/issues/19974. We can extraction of the MGF and passing that to `DecryptOAEP` when upping the Go version of this library.

Reply:
Ok that was added in Go 1.20